File: procedural/Scala/caesar_cipher.scala
object CaesarCipher {
    def encrypt(toCipher: String, shift: Int): String = {
		var ciphered: String = "";
        for(i <- 0 to (toCipher.length() - 1)) {
                // Account for uppercase letters
                if (toCipher.charAt(i) >= 'A' && toCipher.charAt(i) <= 'Z') {
                    // If no whitespace, shift character
                    if(!((toCipher.charAt(i)).isWhitespace)) {
                        ciphered = ciphered + (((toCipher.charAt(i).toInt + shift - 65) % 26).toInt + 65).toChar.toString();
                    }
                    // Else whitespace, append space 
                    else {
                        ciphered = ciphered + toCipher.charAt(i);
                    }
                }
                // Account for lowercase letters
                else {
                    if(!((toCipher.charAt(i)).isWhitespace)) {
                        ciphered = ciphered + (((toCipher.charAt(i).toInt + shift - 97) % 26).toInt + 97).toChar.toString();
                    }
                    else {
                        ciphered = ciphered + toCipher.charAt(i);
                    }
                }
		}
		return ciphered;
	}


    def decrypt(ciphered: String, shift: Int): String = {
        // Shifts back to original
        var negativeShift: Int = (26 - shift)
        return encrypt(ciphered, negativeShift)
    }


    def solve(cipherStr: String, loop: Int): Unit = {
        for(i <- 0 to loop) {
            println("Caesar " + i + ": " + encrypt(cipherStr, i))
        }
    }
    
    def main(args: Array[String]): Unit = {
        
        // Initialization
        var cipherStr = "Dude Wheres My Car"
        var shift = 4
        
        println("Welcome to Caesar Ciphers with Scala!")
        
        // Call Caesar Cipher Encryption Method
        var ciphered = encrypt(cipherStr, shift)
        println("Encrypted Cipher Value: " + ciphered)
  
        // Call Caesar Cipher Decryption Method
        var deciphered = decrypt(ciphered, shift)
        println("Decrypted Cipher Value: " + deciphered)
      
        // Call Caesar Cipher Solve Method
        solve("DUDE", 26)
    }
}

File: procedural/Fortran/caesar_cipher.f90
program Caesar_Cipher
  implicit none
 
  ! Initialization
  character(43) :: cipherStr = "Dude Wheres My Car"
  integer :: shift = 4  
 
  write(*, "(2a)") "Welcome to Caesar Ciphers with Fortran!"
  
  ! Call Caesar Cipher Encryption Method
  call encrypt(cipherStr, shift)
  write(*, "(2a)") "Encrypted Cipher Value: ", cipherStr
  
  ! Call Caesar Cipher Decryption Method
  call decrypt(cipherStr, shift)
  write(*, "(2a)") "Decrypted Cipher Value: ", cipherStr
  
  ! Call Caesar Cipher Solve Method
  call solve("DUDE", 26)
 
contains

 
subroutine encrypt(toCipher, shift)
  implicit none
  character(*) :: toCipher
  integer :: shift
  integer :: i
 
  do i = 1, len(toCipher)
    select case(toCipher(i:i))
      ! Account for uppercase letters
      case ('A':'Z')
        toCipher(i:i) = achar(modulo(iachar(toCipher(i:i)) - 65 + shift, 26) + 65)
      ! Account for lowercase letters
      case ('a':'z')
        toCipher(i:i) = achar(modulo(iachar(toCipher(i:i)) - 97 + shift, 26) + 97)
    end select
  end do
end subroutine
 
 
subroutine decrypt(ciphered, shift)
  implicit none
  character(*) :: ciphered
  integer :: shift
  integer :: reverse
  
  reverse = (shift * (-1))
  call encrypt(ciphered, reverse)
  
end subroutine


subroutine solve(cipher, loop)
  implicit none
  character(4) :: cipher
  integer :: loop
  character(4) :: newCiphered
  integer :: i
  
  do i = 0, loop
    ! Breaks without reassign
    newCiphered = cipher
    call encrypt(newCiphered, i)
    write(*, *) "Caesar ", i, ": ", newCiphered
  end do
  
end subroutine


end program Caesar_Cipher